Feature permissions compact

flbulgarelli opened the pull request:
# :dart:  Goal

To compact permissions when parsing and updating

# :memo: Details

This PR refactors the role objects, to make them a bit more robust. It also removes narrower permissions when wider ones are added. Conversely, it avoids adding narrower permissions when wider permissions already exist. 

I used a similar design to the grants `push_and_compact!` method, which allow me to reuse a bunch of private methods.

flbulgarelli committed:
Add superseded_by?

flbulgarelli committed:
Include compact logic into permissions

flbulgarelli committed:
Add test for wider grants scenarios

flbulgarelli committed:
Fix rspec warnings

flbulgarelli committed:
Always compact permission's roles

flbulgarelli committed:
Make promotion to Role ojects more robust

flbulgarelli committed:
Fix code style

julian-berbel commented on lib/mumukit/auth/permissions.rb:
```diff
@@ -54,7 +54,13 @@ def granted_organizations_for(role)
 
   def add_permission!(role, *grants)
-    scope_for(role).add_grant! *grants
+    role = role.to_mumukit_role
```
either this or line #156 should not be necessary, right?

flbulgarelli replied:
No, it is intended, since `push_and_compact!` may be called from other places. On the other hand, calling this here avoids re-running the conversion within the `each` loop - since it will be a no-op.

julian-berbel commented on lib/mumukit/auth/role.rb:
```diff
@@ -17,15 +30,33 @@ def to_sym
 
-    def self.parent(parent)
-      define_method(:parent) { self.class.parse(parent) }
+    def narrower_than_other?(other)
```
Couldn't this recurse over `narrower_than_other?` instead of `narrower_than?`
though it's just a few less comparisons, not a big deal

flbulgarelli replied:
You are right. It has the drawback it can not be truly private, but I will change it.

julian-berbel commented on lib/mumukit/auth/role.rb:
```diff
@@ -64,6 +95,10 @@ class Owner < Role
         false
       end
+
+      def narrower_than_other?(*)
```
to keep consistency this should be private

flbulgarelli replied:
You are right!

flbulgarelli committed:
Improve empty? in permissions

flbulgarelli committed:
Make _narrower_than_other consistently private

Implement @julian-berbel suggestions. Instead of making it truly private
it is consistently made private by convention in order to allow it to be
recursed

flbulgarelli commented:
:up: This PR was not-so compatible as supposed. Not a big deal, but care should be taken when updating code
